File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        ans = []
        nums.sort()
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:  continue
            j = i+1
            k = len(nums) -1
            tar = -nums[i]
            while j < k:
                sum = nums[j] + nums[k]
                if sum == tar:
                    ans.append([nums[i], nums[j], nums[k]])
                    while j < k and nums[j+1] == nums[j]:
                        j +=1
                    while j < k and nums[k-1] == nums[k]:
                        k -= 1
                    j +=1
                    k -=1
                elif sum < tar:
                    j +=1
                else:
                    k -= 1
            
        return ans

File: 15-3sum/3sum.cpp
typedef vector<vector<int>> vvi;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vvi ans;
        sort(nums.begin(), nums.end());
        for(int i = 0; i<nums.size()-2; ++i){
            if(i>0 && nums[i] == nums[i-1])   continue;
            int tar = - nums[i];
            int j = i+1, k = nums.size() - 1;
            while(j < k){
                int sum = nums[j] + nums[k];
                if(sum == tar){
                    ans.push_back({nums[i], nums[j], nums[k]});
                    while(j<k && nums[j+1] == nums[j])    j++;
                    while(k>j && nums[k-1] == nums[k])    k--;
                    j++; k--;
                }
                else if(sum < tar)
                    j++;
                else
                    k--;
            }
        }
        return ans;
    }
};